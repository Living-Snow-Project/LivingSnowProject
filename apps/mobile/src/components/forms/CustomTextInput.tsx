import React, { forwardRef /* useRef */ } from "react";
import { FormControl, Input, WarningOutlineIcon } from "native-base";
/* import {
  NativeSyntheticEvent,
  TextInputContentSizeChangeEventData,
} from "react-native";
import RCTDeviceEventEmitter from "react-native/Libraries/EventEmitter/RCTDeviceEventEmitter"; */

type CustomTextInputProps = {
  label: string;
  placeholder: string;
  value?: string | undefined;
  validation?: string;
  maxLength?: number;
  isDisabled?: boolean;
  isInvalid?: boolean;
  isRequired?: boolean;
  onChangeText?: (text: string) => void;
  onSubmitEditing?: () => void;
};

export const CustomTextInput = forwardRef<typeof Input, CustomTextInputProps>(
  (
    {
      label,
      placeholder,
      value,
      validation,
      maxLength = 255,
      isDisabled = false,
      isInvalid = false,
      isRequired = false,
      onChangeText = () => {},
      onSubmitEditing = () => {},
    }: CustomTextInputProps,
    ref
  ) => (
    /* const textInputHeight = useRef(0);

    // onContentSizeChange is called frequently for multiline TextInput, we only want to emit 'keyboardDidShow' event when height actually changes
    const handleMultilineTextInputOnContentSizeChange = (
      height: number,
      event: NativeSyntheticEvent<TextInputContentSizeChangeEventData>
    ) => {
      if (height !== 0 && height !== event.nativeEvent.contentSize.height) {
        // 'keyboardDidShow' expects the height of the keyboard (which we could capture in a new event listener in this component)
        // since we only have 1 'keyboardDidShow' listener we changed its logic to respond to this input
        // this is potentially bad if we set up additional 'keyboardDidShow' listeners in the app
        RCTDeviceEventEmitter.emit("keyboardDidShow", {});
      }

      return event.nativeEvent.contentSize.height;
    }; */

    <FormControl isRequired={isRequired} isInvalid={isInvalid}>
      <FormControl.Label>{label}</FormControl.Label>
      <Input
        size="lg"
        multiline
        blurOnSubmit
        // @ts-ignore ref typing isn't quite right in native-base yet
        ref={ref}
        value={value}
        placeholder={placeholder}
        isDisabled={isDisabled}
        onChangeText={onChangeText}
        onSubmitEditing={onSubmitEditing}
        /* onContentSizeChange={(event) => {
          textInputHeight.current =
            handleMultilineTextInputOnContentSizeChange(
              textInputHeight.current,
              event
            );
        }} */
        maxLength={maxLength}
        returnKeyType="done"
        variant="outline"
      />
      <FormControl.ErrorMessage leftIcon={<WarningOutlineIcon size="sm" />}>
        {validation}
      </FormControl.ErrorMessage>
    </FormControl>
  )
);

// this is needed because of forwardRef
CustomTextInput.displayName = `CustomTextInput`;
