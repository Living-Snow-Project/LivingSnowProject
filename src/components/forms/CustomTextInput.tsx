import React, { forwardRef, useRef } from "react";
import { Keyboard, Platform, Text, TextInput } from "react-native";
import PropTypes from "prop-types";
import { formInputStyles } from "../../styles/FormInput";

type CustomTextInputProps = {
  description: string;
  placeholder: string;
  maxLength?: number;
  onChangeText: (text: string) => void;
  onSubmitEditing?: () => void;
};

const CustomTextInput = forwardRef<TextInput, CustomTextInputProps>(
  (
    {
      description,
      placeholder,
      maxLength = 255,
      onChangeText,
      onSubmitEditing = () => {},
    }: CustomTextInputProps,
    ref
  ) => {
    const textInputHeight = useRef(0);

    // onContentSizeChange is called frequently for multiline TextInput, we only want to emit 'keyboardDidShow' event when height actually changes
    const handleMultilineTextInputOnContentSizeChange = (height, event) => {
      if (height !== 0 && height !== event.nativeEvent.contentSize.height) {
        // 'keyboardDidShow' expects the height of the keyboard (which we could capture in a new event listener in this component)
        // since we only have 1 'keyboardDidShow' listener we changed its logic to respond to this input
        // this is potentially bad if we set up additional 'keyboardDidShow' listeners in the app
        Keyboard.emit("keyboardDidShow", {});
      }

      return event.nativeEvent.contentSize.height;
    };

    return (
      <>
        <Text style={formInputStyles.optionStaticText}>{description}</Text>
        <TextInput
          // this is kind of a hack to get placeholder text to appear vertically centered on ios when TextInput is multiline
          style={[
            formInputStyles.optionInputText,
            Platform.OS === "ios" ? formInputStyles.multilineTextInput : {},
          ]}
          multiline
          blurOnSubmit
          ref={ref}
          placeholder={placeholder}
          onChangeText={(text) => onChangeText(text)}
          onSubmitEditing={() => onSubmitEditing()}
          onContentSizeChange={(event) => {
            textInputHeight.current =
              handleMultilineTextInputOnContentSizeChange(
                textInputHeight.current,
                event
              );
          }}
          maxLength={maxLength}
          returnKeyType="done"
        />
      </>
    );
  }
);

CustomTextInput.propTypes = {
  description: PropTypes.string.isRequired,
  placeholder: PropTypes.string.isRequired,
  maxLength: PropTypes.number,
  onChangeText: PropTypes.func.isRequired,
  onSubmitEditing: PropTypes.func,
};

CustomTextInput.defaultProps = {
  maxLength: 255,
  onSubmitEditing: () => {},
};

CustomTextInput.displayName = `CustomTextInput`;

export default CustomTextInput;
