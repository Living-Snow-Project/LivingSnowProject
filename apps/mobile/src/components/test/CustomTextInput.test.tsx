import React from "react";
import { Keyboard, Platform } from "react-native";
import { ReactTestInstance } from "react-test-renderer";
import { render, fireEvent } from "@testing-library/react-native";
import CustomTextInput from "../forms/CustomTextInput";

describe("CustomTextInput tests", () => {
  const props = {
    description: "test custom text input",
    placeholder: "custom text input placeholder",
    onChangeText: jest.fn(),
  };
  let renderer;
  let customTextInput: ReactTestInstance;

  beforeEach(() => {
    renderer = render(<CustomTextInput {...props} />);

    customTextInput = renderer.getByPlaceholderText(props.placeholder);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  // additional paddingTop in styles array
  test("renders on ios", () => {
    const { toJSON } = renderer;
    expect(toJSON()).toMatchSnapshot();
  });

  test("renders on android", () => {
    Platform.OS = "android";
    const { toJSON } = render(<CustomTextInput {...props} />);
    expect(toJSON()).toMatchSnapshot();
    Platform.OS = "ios";
  });

  test("change text", () => {
    const testText = "test text";
    const { getByDisplayValue } = renderer;

    expect(customTextInput).not.toBeNull();
    fireEvent.changeText(customTextInput, testText);
    expect(getByDisplayValue(testText)).not.toBeNull();
    expect(props.onChangeText).toHaveBeenCalledWith(testText);
  });

  test("multiline text", () => {
    const keyboardListener = jest.fn();

    Keyboard.addListener("keyboardDidShow", keyboardListener);

    // could not find an end user way to fire keyboard events in React Native
    // these implementation details simulate the multiline text input growing vertically
    // userEvents and testing-dom libraries currently only work browser environment
    fireEvent(customTextInput, "onContentSizeChange", {
      nativeEvent: { contentSize: { height: 30 } },
    });

    fireEvent(customTextInput, "onContentSizeChange", {
      nativeEvent: { contentSize: { height: 60 } },
    });

    expect(keyboardListener).toBeCalledTimes(1);
  });

  test("submit text", () => {
    const onSubmitEditing = jest.fn();
    const { getByPlaceholderText } = render(
      <CustomTextInput
        description="test submitting"
        placeholder={props.placeholder}
        maxLength={10}
        onChangeText={() => {}}
        onSubmitEditing={onSubmitEditing}
      />
    );

    fireEvent(getByPlaceholderText(props.placeholder), "onSubmitEditing");
    expect(onSubmitEditing).toBeCalledTimes(1);
  });
});
