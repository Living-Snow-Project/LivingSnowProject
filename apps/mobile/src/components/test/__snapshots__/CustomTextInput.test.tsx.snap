// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomTextInput tests renders on android 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    dataSet={{}}
    isInvalid={false}
    isReadOnly={false}
    isRequired={false}
    style={
      {
        "width": "100%",
      }
    }
  >
    <View
      dataSet={{}}
      feedbackId="field-8-feedback"
      hasFeedbackText={true}
      hasHelpText={false}
      helpTextId="field-8-helptext"
      isDisabled={false}
      isInvalid={false}
      isReadOnly={false}
      isRequired={false}
      labelId="field-8-label"
      nativeID="field-8-label"
      setHasFeedbackText={[Function]}
      setHasHelpText={[Function]}
      style={
        {
          "flexDirection": "row",
          "justifyContent": "flex-start",
          "marginBottom": 4,
          "marginTop": 4,
        }
      }
    >
      <Text
        dataSet={{}}
        style={
          {
            "backgroundColor": undefined,
            "color": "#737373",
            "fontFamily": undefined,
            "fontSize": 14,
            "fontStyle": "normal",
            "fontWeight": "500",
            "letterSpacing": 0,
            "lineHeight": 21,
            "textDecorationLine": undefined,
          }
        }
      >
        test custom text input
      </Text>
    </View>
    <View
      dataSet={{}}
      isFocused={false}
      style={
        {
          "alignItems": "center",
          "borderColor": "#d4d4d4",
          "borderRadius": 4,
          "borderWidth": 1,
          "flexDirection": "row",
          "overflow": "hidden",
        }
      }
    >
      <TextInput
        accessibilityDescribedBy="field-8-feedback"
        accessible={true}
        blurOnSubmit={false}
        dataSet={{}}
        disabled={false}
        editable={true}
        isRequired={false}
        maxLength={255}
        nativeID="field-8-input"
        onBlur={[Function]}
        onChangeText={[MockFunction]}
        onFocus={[Function]}
        onKeyPress={[Function]}
        onSubmitEditing={[Function]}
        placeholder="custom text input placeholder"
        placeholderTextColor="#a3a3a3"
        readOnly={false}
        required={false}
        returnKeyType="done"
        secureTextEntry={false}
        style={
          {
            "backgroundColor": "transparent",
            "color": "#171717",
            "flex": 1,
            "fontFamily": undefined,
            "fontSize": 16,
            "fontStyle": "normal",
            "fontWeight": "400",
            "height": "100%",
            "paddingBottom": 8,
            "paddingLeft": 12,
            "paddingRight": 12,
            "paddingTop": 8,
            "width": "100%",
          }
        }
      />
    </View>
  </View>
</RNCSafeAreaProvider>
`;

exports[`CustomTextInput tests renders on ios 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    dataSet={{}}
    isInvalid={false}
    isReadOnly={false}
    isRequired={false}
    style={
      {
        "width": "100%",
      }
    }
  >
    <View
      dataSet={{}}
      feedbackId="field-2-feedback"
      hasFeedbackText={true}
      hasHelpText={false}
      helpTextId="field-2-helptext"
      isDisabled={false}
      isInvalid={false}
      isReadOnly={false}
      isRequired={false}
      labelId="field-2-label"
      nativeID="field-2-label"
      setHasFeedbackText={[Function]}
      setHasHelpText={[Function]}
      style={
        {
          "flexDirection": "row",
          "justifyContent": "flex-start",
          "marginBottom": 4,
          "marginTop": 4,
        }
      }
    >
      <Text
        dataSet={{}}
        style={
          {
            "backgroundColor": undefined,
            "color": "#737373",
            "fontFamily": undefined,
            "fontSize": 14,
            "fontStyle": "normal",
            "fontWeight": "500",
            "letterSpacing": 0,
            "lineHeight": 21,
            "textDecorationLine": undefined,
          }
        }
      >
        test custom text input
      </Text>
    </View>
    <View
      dataSet={{}}
      isFocused={false}
      style={
        {
          "alignItems": "center",
          "borderColor": "#d4d4d4",
          "borderRadius": 4,
          "borderWidth": 1,
          "flexDirection": "row",
          "overflow": "hidden",
        }
      }
    >
      <TextInput
        accessibilityDescribedBy="field-2-feedback"
        accessible={true}
        blurOnSubmit={false}
        dataSet={{}}
        disabled={false}
        editable={true}
        isRequired={false}
        maxLength={255}
        nativeID="field-2-input"
        onBlur={[Function]}
        onChangeText={[MockFunction]}
        onFocus={[Function]}
        onKeyPress={[Function]}
        onSubmitEditing={[Function]}
        placeholder="custom text input placeholder"
        placeholderTextColor="#a3a3a3"
        readOnly={false}
        required={false}
        returnKeyType="done"
        secureTextEntry={false}
        style={
          {
            "backgroundColor": "transparent",
            "color": "#171717",
            "flex": 1,
            "fontFamily": undefined,
            "fontSize": 16,
            "fontStyle": "normal",
            "fontWeight": "400",
            "height": "100%",
            "paddingBottom": 8,
            "paddingLeft": 12,
            "paddingRight": 12,
            "paddingTop": 8,
            "width": "100%",
          }
        }
      />
    </View>
  </View>
</RNCSafeAreaProvider>
`;
